src/dataframe/dataframe.jl

```diff
@@ -216,6 +215,21 @@ struct DataFrame <: AbstractDataFrame
     end
 end
 
+function _preprocess_column(col, len, copycols)
+    # check for vectors first as they are most common
+    col isa AbstractRange && return collect(col)
+    col isa AbstractVector && return copycols ? copy(col) : col
+    if col isa Union{AbstractArray{<:Any, 0}, Ref}
+        x = col[]
+        return fill!(Tables.allocatecolumn(typeof(x), len), x)
+    end
+    if col isa AbstractArray
+        throw(ArgumentError("adding AbstractArray other than AbstractVector " *
+                            "as a column of a data frame is not allowed"))
+    end
```

Comment:
Check this first? Or how about keeping the original `elseif` structure for clarity?

Reply:
I have to check it only at this stage. I have reverted the `elseif` logic